### PR TITLE
fix(async_inference): do not resize camera frames to the dataset resolution

### LINKED ISSUE #2475: Why there is difference between async inference and local inference in image resize?
I read code between `src/lerobot/async_inference/policy_server.py` and `src/lerobot/scripts/lerobot_record.py`. I found difference in these 2 code about inference which causes different image shape
1. `src/lerobot/scripts/lerobot_record.py` use this to deal with observation
And `prepare_observation_for_inference` is like this:
```python
def prepare_observation_for_inference(
    observation: dict[str, np.ndarray],
    device: torch.device,
    task: str | None = None,
    robot_type: str | None = None,
) -> RobotObservation:
    for name in observation:
        observation[name] = torch.from_numpy(observation[name])
        if "image" in name:
            observation[name] = observation[name].type(torch.float32) / 255
            observation[name] = observation[name].permute(2, 0, 1).contiguous()
        observation[name] = observation[name].unsqueeze(0)
        observation[name] = observation[name].to(device)
    observation["task"] = task if task else ""
    observation["robot_type"] = robot_type if robot_type else ""
    return observation
```

Here no **resize** operation in images i think. 

2. in async_inference policy_server.py,it uses
But here function`prepare_raw_observation` makes sense on image shape
```python
def prepare_raw_observation(
    robot_obs: RawObservation,
    lerobot_features: dict[str, dict],
    policy_image_features: dict[str, PolicyFeature],
) -> Observation:
    """Matches keys from the raw robot_obs dict to the keys expected by a given policy (passed as
    policy_image_features)."""
    # 1. {motor.pos1:value1, motor.pos2:value2, ..., laptop:np.ndarray} ->
    # -> {observation.state:[value1,value2,...], observation.images.laptop:np.ndarray}
    lerobot_obs = make_lerobot_observation(robot_obs, lerobot_features)
    # 2. Greps all observation.images.<> keys
    image_keys = list(filter(is_image_key, lerobot_obs))
    # state's shape is expected as (B, state_dim)
    state_dict = {OBS_STATE: extract_state_from_raw_observation(lerobot_obs)}
    image_dict = {
        image_k: extract_images_from_raw_observation(lerobot_obs, image_k) for image_k in image_keys
    }
    # Turns the image features to (C, H, W) with H, W matching the policy image features.
    # This reduces the resolution of the images
    image_dict = {
        key: resize_robot_observation_image(torch.tensor(lerobot_obs[key]), policy_image_features[key].shape)
        for key in image_keys
    }
    if "task" in robot_obs:
        state_dict["task"] = robot_obs["task"]
    return {**state_dict, **image_dict}
```
Here the shape of observation images is modified to policy config
```python
def resize_robot_observation_image(image: torch.tensor, resize_dims: tuple[int, int, int]) -> torch.tensor:
    assert image.ndim == 3, f"Image must be (C, H, W)! Received {image.shape}"
    # (H, W, C) -> (C, H, W) for resizing from robot obsevation resolution to policy image resolution
    image = image.permute(2, 0, 1)
    dims = (resize_dims[1], resize_dims[2])
    # Add batch dimension for interpolate: (C, H, W) -> (1, C, H, W)
    image_batched = image.unsqueeze(0)
    # Interpolate and remove batch dimension: (1, C, H, W) -> (C, H, W)
    resized = torch.nn.functional.interpolate(image_batched, size=dims, mode="bilinear", align_corners=False)

    return resized.squeeze(0)
``` 

I found this when I can inference correctly locally by made weird action outputs from async inference. it must be caused by my didn't resize input image when training. -.-

version:deb9596bd3796c03ae3a5a6b81b63c1dba296256

### LINKED ISSUE #2980: Async inference behaves differently between migrated older models and newly trained models
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
- LeRobot: commit at 14a15f90
- Robot: SO-101
- Policy: SmolVLA
- OS: macOS

- the output of `lerobot-info`

- LeRobot version: 0.4.4
- Platform: macOS-15.7.2-arm64-arm-64bit
- Python version: 3.11.9
- Huggingface Hub version: 0.35.3
- Datasets version: 4.1.1
- Numpy version: 2.2.6
- FFmpeg version: 7.1.1
- PyTorch version: 2.7.1
- Is PyTorch built with CUDA support?: False
- Cuda version: N/A
- GPU model: N/A
- Using GPU in script?: <fill in>
- lerobot scripts: ['lerobot-calibrate', 'lerobot-dataset-viz', 'lerobot-edit-dataset', 'lerobot-eval', 'lerobot-find-cameras', 'lerobot-find-joint-limits', 'lerobot-find-port', 'lerobot-imgtransform-viz', 'lerobot-info', 'lerobot-record', 'lerobot-replay', 'lerobot-setup-can', 'lerobot-setup-motors', 'lerobot-teleoperate', 'lerobot-train', 'lerobot-train-tokenizer']
```

### Description

We observe a noticeable difference in task success between Sync (local) inference and Async inference.

In our setup, policies trained with the current version of LeRobot perform reliably under Sync inference, but often fail under Async inference. In contrast, policies trained with older LeRobot versions and then migrated to the current format tend to show much smaller degradation when run with Async inference.

The issue is reproducible across multiple runs of the same task and does not appear to be limited to a single dataset or model checkpoint.

#### Observed behavior

* Sync inference achieves relatively stable task success.
* Async inference shows significantly lower success rates.
* The robot arm usually moves close to the target, but precise actions (e.g. grasping) often fail under Async inference.

A consistent difference we observed is how the policy was trained:

* Policies trained with older LeRobot versions (around commit 483be9a ) tend to retain reasonable performance under Async inference.
* Policies fine-tuned from the same base model using the current version of LeRobot often show severe degradation under Async inference, even when trained on the same dataset migrated to Dataset v3.

#### Policies and datasets used

| Policy         | Training setup                               | Dataset version   | Link                                                                                                                                                                       |
| -------------- | -------------------------------------------- | ----------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Old (migrated) | Fine-tuned with older LeRobot, then migrated | Dataset v2.1      | [https://huggingface.co/yutaro-kimura-acs/so101_pp_blue_box_policy_01_migrated](https://huggingface.co/yutaro-kimura-acs/so101_pp_blue_box_policy_01_migrated)             |
| New            | Fine-tuned with current LeRobot              | Dataset v2.1 → v3 (Migrated with [conversion script](https://huggingface.co/docs/lerobot/lerobot-dataset-v3#migrate-v21--v30)) | [https://huggingface.co/aShunSasaki/so101_pp_blue_box_yk_dataset_policy_v30_cp_20000](https://huggingface.co/aShunSasaki/so101_pp_blue_box_yk_dataset_policy_v30_cp_20000) |

Both policies were fine-tuned from the same base model (`smolvla_base`) using the same dataset:
[https://huggingface.co/datasets/yutaro-kimura-acs/so101_pp_blue_box](https://huggingface.co/datasets/yutaro-kimura-acs/so101_pp_blue_box)

#### Videos

* Old (migrated):

https://github.com/user-attachments/assets/d362ed09-112c-46ec-adec-7058c4b6ea3f

* New:

https://github.com/user-attachments/assets/c3c6fcd8-f169-4324-8aba-9148af304e41

#### Task success rates

| Policy         | Sync inference | Async inference | Notes                                    |
| -------------- | :------------: | :-------------: | ---------------------------------------- |
| Old (migrated) |     80–90%     |       **~50%**      | `input_features` shape: `[3, 720, 1280]` |
| New (current)  |     80–90%     |        **0%**       | `input_features` shape: `[3, 256, 256]`  |

* Robot: SO-101
* Task: Pick-and-place (blue box)
* Policy architecture: SmolVLA (`smolvla_base` fine-tuned)

#### Possible contributing factor

One possible factor we identified is a difference in image preprocessing between Sync and Async inference paths.

In the async path, the camera image appears to be resized to match `policy.config.input_features` before being passed to the model. For policies where `input_features` reflects the dataset storage resolution (e.g. 256×256), this results in the camera image being resized from its native resolution (1280×720) to a square shape.

In contrast, Sync inference does not resize the image at this stage and relies on the model’s internal `resize_with_pad`, which preserves aspect ratio.

This behavior seems related to the image resizing discrepancy discussed in #2475, and may help explain why older migrated policies behave more robustly under Async inference.

#### Configuration differences observed

While investigating this issue, we also compared the policy configuration files between older migrated policies and newly fine-tuned ones.

One notable difference is the `input_features` image shape:

* Older migrated policies use the camera’s native resolution (e.g. `[3, 720, 1280]`)
* Newly fine-tuned policies use the dataset storage resolution (e.g. `[3, 256, 256]`)

This difference appears to affect how images are handled in the async inference path, where the input image is resized to match `input_features` before being passed to the policy.

We are not fully sure whether this alone explains the observed behavior, but it stood out during the comparison.

### Relevant logs or stack trace

```Shell
No explicit runtime errors or stack traces were observed.
```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

This issue may be related to the image preprocessing difference discussed in:

- https://github.com/huggingface/lerobot/issues/2475

That issue focuses on image resizing behavior specifically, while this ticket aims to describe the broader performance gap observed between Sync and Async inference, as well as the difference between older migrated models and newly trained ones.

At the moment, using a policy trained with an older LeRobot version and migrating it to the current format appears to mitigate the issue.

### PR DESCRIPTION
**Pull request by aShunSasaki**
## Summary / Motivation

Async inference resized camera frames to `policy.config.image_features` before passing them to the policy. This shape comes from the resolution used when recording the training dataset and is not necessarily related to the camera resolution used during inference. The resize also changed the aspect ratio because it used plain bilinear interpolation. This interfered with the aspect ratio preserving resize inside each policy, such as `resize_with_pad` and `resize_with_pad_torch`.

For pi0, pi05, and pi0_fast, `image_resolution` is square. Resizing to that shape before policy preprocessing caused the policy's internal resize to be skipped completely.

This issue was reported in #2980. With the same hardware and task, a SmolVLA checkpoint fine tuned on the current dataset format achieved an 80 to 90 percent success rate with sync inference and approximately 0 percent with async inference.

The resize was performed by:

`prepare_raw_observation()` → `resize_robot_observation_image()` in `helpers.py`

Sync inference does not resize camera frames before passing them to the policy. This behavior existed only in async inference.

I measured the output against sync inference using the same configuration as #2980:

* SmolVLA
* 1280x720 camera
* `input_features=(3,256,256)`
* `resize_imgs_with_padding=(512,512)`

| Implementation | Result against sync inference |
|---|---:|
| Current `main` | 8.38 dB PSNR |
| #3158 with center padding | 7.95 dB PSNR |
| This PR | Bit identical |

## Related issues

* Fixes #2980
* Fixes #2475
* Supersedes #3158. Thanks to @jashshah999 for the earlier attempt.
* Related to #3832

## What changed

* Camera frames are now passed to the policy at their native resolution, matching sync inference.
* Removed `resize_robot_observation_image()`.
* Removed the unused `policy_image_features` argument from `prepare_raw_observation`.
* Removed the unused `policy_image_features` argument from `raw_observation_to_observation`.
* Removed `PolicyServer.policy_image_features`.
* Added validation that raises `ValueError` when a camera frame does not have the expected `(H, W, C)` shape.
* This does not introduce a user facing breaking change.

## How was this tested (or how to run locally)

* Ran `pre-commit run -a`.
* Ran `pytest tests/async_inference`.
* Added `test_sync_and_async_preprocessing_are_identical`.
* Tested 480p, 720p, and mixed resolution camera inputs.
* Verified that `build_inference_frame` for sync inference and `raw_observation_to_observation` for async inference produce bit identical tensors from the same robot observation.
* Verified that the new parity test fails against `main` because the old function signature still requires the resize argument, and passes with this change.
* Updated existing tests for the native, unresized camera frame shapes.
* Added a test that verifies an invalid camera frame shape raises `ValueError`.

## Checklist (required before merge)

* [x] Linting/formatting run (`pre-commit run -a`)
* [x] All tests pass locally (`pytest`)
  * The async inference test suite passes with `pytest tests/async_inference`.
* [x] Documentation updated
  * No documentation update is required because this does not change the public API or user workflow.
* [ ] CI is green
* [x] Community Review: I reviewed #3158 and left feedback there.

## Reviewer notes

Please focus on whether camera preprocessing should remain fully owned by each policy.

I also checked pi0 and pi05. When `helpers.py` resizes the image first, their internal resize becomes a no op. The padding method selected in `helpers.py` therefore becomes the final preprocessing result.

The required behavior differs between policy families:

* SmolVLA and XVLA use top left padding.
* pi0, pi05, and pi0_fast use center padding.
* GR00T uses shortest edge resizing and cropping.
* EVO1 and MolmoAct2 do not perform an internal resize.

Because of these differences, there is no single resize or padding rule in `helpers.py` that is correct for every policy.

There is no additional network cost. The previous resize happened on the server in `PolicyServer._predict_action_chunk`, after the observation had already been received over gRPC.

The additional server CPU cost is approximately 0.1 to 1.2 ms per frame, depending on the camera resolution. This is small compared with the policy forward time.

AI disclosure: I used Claude Code for the investigation, implementation, tests, and numeric verification, including the PSNR and parity measurements. I reviewed the code, test results, and measurements myself.

### CODE REVIEW

File: src/lerobot/async_inference/helpers.py [hunk 1]
```diff
@@ -128,9 +114,21 @@ def extract_state_from_raw_observation(
...
+    Only the (H, W, C) -> (C, H, W) transposition is applied: the camera's native resolution is
+    preserved, so the policy's own aspect-ratio-preserving resize (e.g. `resize_with_pad`) sees
+    exactly the same input it would see during sync inference. Resizing here to
+    `policy.config.image_features` -- the resolution the training *dataset* was recorded at, which
+    has nothing to do with the camera resolution used at inference time -- used to distort the
+    aspect ratio and short-circuit that internal resize. See #2980.
+    """
+    image = torch.tensor(lerobot_obs[camera_key])
+    if image.ndim != 3:
+        raise ValueError(f"Camera image must be (H, W, C)! Received {tuple(image.shape)} for {camera_key}")
```
[the remark this answers is not in the document]
Comment (aShunSasaki) [reply src/lerobot/async_inference/helpers.py]: I fixed it in the latest commit.

`extract_images_from_raw_observation()` now checks that `image.shape[-1]` is either `1` or `3`. This matches the channel last format already used by `hw_to_dataset_features` elsewhere in the codebase.

I also added a regression test for the `(3, 480, 640)` case. I confirmed that the test fails with `DID NOT RAISE ValueError` before the fix and passes with the latest change.
